core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -749,10 +748,16 @@ class Partition(val topicPartition: TopicPartition,
    * since all callers of this private API acquire that lock
    */
   private def maybeIncrementLeaderHW(leaderLog: Log, curTime: Long = time.milliseconds): Boolean = {
```

Comment:
Leaving this here for lack of an alternative. I think for low-volume topics, we are actually hitting this logic even when the fetch offset hasn't changed. But if the fetch offset hasn't changed, then there should be no need to expand the ISR or to increment the high watermark. So what I was thinking is that `Replica.updateFetchState` could return a flag indicating whether or not the follower's end offset has changed. If it hasn't, we skip the ISR expansion check. What do you think?